model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,849 @@
+        ModelCriteriaBuilder<UserLoginFailureModel> mcb = userLoginFailureStore.createCriteriaBuilder()
+                .compare(UserLoginFailureModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserLoginFailureModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, userId);
+
+        LOG.tracef("getUserLoginFailure(%s, %s)%s", realm, userId, getShortStackTrace());
+
+        return userLoginFailureTx.getUpdatedNotRemoved(mcb)
+                .map(userLoginFailureEntityToAdapterFunc(realm))
+                .findFirst()
+                .orElse(null);
+    }
+
+    @Override
+    public UserLoginFailureModel addUserLoginFailure(RealmModel realm, String userId) {
+        ModelCriteriaBuilder<UserLoginFailureModel> mcb = userLoginFailureStore.createCriteriaBuilder()
+                .compare(UserLoginFailureModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserLoginFailureModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, userId);
+
+        LOG.tracef("addUserLoginFailure(%s, %s)%s", realm, userId, getShortStackTrace());
+
+        MapUserLoginFailureEntity userLoginFailureEntity = userLoginFailureTx.getUpdatedNotRemoved(mcb).findFirst().orElse(null);
+
+        if (userLoginFailureEntity == null) {
+            userLoginFailureEntity = new MapUserLoginFailureEntity(UUID.randomUUID(), realm.getId(), userId);
+
+            userLoginFailureTx.create(userLoginFailureEntity.getId(), userLoginFailureEntity);
+        }
+
+        return userLoginFailureEntityToAdapterFunc(realm).apply(userLoginFailureEntity);
+    }
+
+    @Override
+    public void removeUserLoginFailure(RealmModel realm, String userId) {
+        ModelCriteriaBuilder<UserLoginFailureModel> mcb = userLoginFailureStore.createCriteriaBuilder()
+                .compare(UserLoginFailureModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserLoginFailureModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, userId);
+
+        LOG.tracef("removeUserLoginFailure(%s, %s)%s", realm, userId, getShortStackTrace());
+
+        userLoginFailureTx.delete(UUID.randomUUID(), mcb);
+    }
+
+    @Override
+    public void removeAllUserLoginFailures(RealmModel realm) {
+        ModelCriteriaBuilder<UserLoginFailureModel> mcb = userLoginFailureStore.createCriteriaBuilder()
+                .compare(UserLoginFailureModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId());
+
+        LOG.tracef("removeAllUserLoginFailures(%s)%s", realm, getShortStackTrace());
+
+        userLoginFailureTx.delete(UUID.randomUUID(), mcb);
+    }
+
+    @Override
+    public void onRealmRemoved(RealmModel realm) {
+        removeUserSessions(realm);
+        removeAllUserLoginFailures(realm);
+    }
+
+    @Override
+    public void onClientRemoved(RealmModel realm, ClientModel client) {
```

Comment:
Client sessions are lazily removed in `MapUserSessionAdapter#getAuthenticatedClientSessions()` when demanded. 
There is no need to delete anything from a DB because this map implementation doesn't use `UserSessionPersisterProvider` at all.